File: frontend/src/types/goal.ts
/**
 * TypeScript types for financial goals.
 */

export interface Goal {
  id: string;
  user_id: string;
  name: string;
  target_amount: number;
  current_amount: number;
  target_date: string | null;
  description: string | null;
  created_at: string;
  updated_at: string;
}

export interface GoalCreate {
  name: string;
  target_amount: number;
  current_amount?: number;
  target_date?: string;
  description?: string;
}

export interface GoalUpdate {
  name?: string;
  target_amount?: number;
  current_amount?: number;
  target_date?: string;
  description?: string;
}

export type GoalStatus = 'Completed' | 'On Track' | 'Behind Target';

export interface GoalStats {
  totalGoals: number;
  activeGoals: number;
  totalTargetValue: number;
  overallProgress: number;
  completedCount: number;
}

File: backend/app/schemas/goal.py
"""
Pydantic schemas for goal endpoints.
"""

import uuid
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class GoalBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    target_amount: float = Field(..., gt=0)
    current_amount: float = Field(0, ge=0)
    target_date: date | None = None
    description: str | None = None


class GoalCreate(GoalBase):
    pass


class GoalUpdate(BaseModel):
    name: str | None = Field(None, min_length=1, max_length=255)
    target_amount: float | None = Field(None, gt=0)
    current_amount: float | None = Field(None, ge=0)
    target_date: date | None = None
    description: str | None = None


class GoalResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    user_id: str
    name: str
    target_amount: float
    current_amount: float
    target_date: date | None
    description: str | None
    created_at: datetime
    updated_at: datetime
